edx_sga/sga.py

```diff
@@ -0,0 +1,530 @@
+        """
+        score = submissions_api.get_score(self.student_submission_id(id))
+        if score:
+            return score['points_earned']
+
+    @reify
+    def score(self):
+        return self.get_score()
+
+    def student_view(self, context=None):
+        """
+        The primary view of the StaffGradedAssignmentXBlock, shown to students
+        when viewing courses.
+        """
+        context = {
+            "student_state": json.dumps(self.student_state()),
+            "id": self.location.name.replace('.', '_')
+        }
+        if self.show_staff_grading_interface():
+            context['is_course_staff'] = True
+            self.update_staff_debug_context(context)
+
+        fragment = Fragment()
+        fragment.add_content(
+            render_template(
+                'templates/staff_graded_assignment/show.html',
+                context
+            )
+        )
+        fragment.add_css(_resource("static/css/edx_sga.css"))
+        fragment.add_javascript(_resource("static/js/src/edx_sga.js"))
+        fragment.initialize_js('StaffGradedAssignmentXBlock')
+        return fragment
+
+    def update_staff_debug_context(self, context):
+        published = self.start
+        context['is_released'] = published and published < _now()
+        context['location'] = self.location
+        context['category'] = type(self).__name__
+        context['fields'] = [
+            (name, field.read_from(self))
+            for name, field in self.fields.items()]
+
+    def student_state(self):
+        """
+        Returns a JSON serializable representation of student's state for
+        rendering in client view.
+        """
+        submission = self.get_submission()
+        if submission:
+            uploaded = {"filename": submission['answer']['filename']}
+        else:
+            uploaded = None
+
+        if self.annotated_sha1:
+            annotated = {"filename": self.annotated_filename}
+        else:
+            annotated = None
+
+        score = self.score
+        if score is not None:
+            graded = {'score': score, 'comment': self.comment}
+        else:
+            graded = None
+
+        return {
+            "uploaded": uploaded,
+            "annotated": annotated,
+            "graded": graded,
+            "max_score": self.max_score(),
+            "upload_allowed": self.upload_allowed(),
+        }
+
+    def staff_grading_data(self):
+        def get_student_data():
+            # Submissions doesn't have API for this, just use model directly
+            students = SubmissionsStudent.objects.filter(
+                course_id=self.course_id,
+                item_id=self.block_id)
+            for student in students:
+                submission = self.get_submission(student.student_id)
+                if not submission:
+                    continue
+                user = user_by_anonymous_id(student.student_id)
+                module, _ = StudentModule.objects.get_or_create(
```

Comment:
An XBlock should never directly access StudentModule models directly. Anything you need from the modulestore or form the StudentModule models should be available through XBlock services, or we should create new XBlock Services to safely expose what's needed.

The username should be available now. An XBlock user service was created recently and could be leveraged here. It appears the module state is being retrieved to view student state from one user to another. You can avoid the need for accessing the state directly by adding the annotated_filename to the submission JSON blob as an arbitrary attribute. This will be useful in debugging anyhow, to have all relative information captured in the submissions table.